enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1123 @@
+  Answer: The connection is still encrypted in public_ssl, so it is not possible to examine or modify the HTTP session there. The fe_sni and fe_no_sni frontends terminate TLS.
+
+#### Can we use conditional values for headers?
+  Answer: No.  You cannot set a conditional value.
+  
+  For example: If you set x-frame-options: "%[res.hdr(server)] if { req.hdr(user-agent) -m sub evil}" on an 
+  HTTP x-frame-options response header, then the response value will be
+  the literal:  `x-frame-options: Apache/2.4.37 (centos) if { req.hdr(user-agent) -m sub evil}`. 
+  
+  In other words, the condition is not evaluated.
+  
+  All values are escaped and quoted using the single quotes. Please check section `Alternative quoting strategies for haproxy.config` for more
+  details. So the value we get after escape and quote we get is a string. 
+
+#### Why is it okay to still allow the host header on the route?
+  Answer: Custom headers that are specified on the route get configured on the corresponding backend, and the backend configuration takes effect after the frontend has done its map lookup.
+
+#### Why do we limit to max 20 items each for custom response and request headers ?
+  Answer: We selected 20 items as this is a max value we can [scale](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#estimated-cost-limits) 
+  w.r.t header value max size 16,384 bytes along with the CEL conditions defined.
+  
+  Setting a non-scalable value results in error [`Forbidden: estimated rule cost exceeds budget by factor of 2.7x (try simplifying the rule, or adding maxItems, maxProperties,
+  and maxLength where arrays, maps, and strings are declared)`](https://github.com/openshift/kubernetes/blob/af29f64acbf9e6d33cde7f3194cf858f54058d3b/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1113)
+  while running integration tests.
+
+#### Why did we select max value size of 16384 for header value ?
+  Answer: This limit was [selected](https://www.geekersdigest.com/max-http-request-header-size-server-comparison/) as most common web servers have a limit of 16384 characters or some lower limit.
+  Note that the total size of all added headers *after* interpolating dynamic values
+  must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
+  IngressController.
+
+#### Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?
+  Answer: It's quite common to have a global setting with an escape hatch locally to override the global setting.
+  The purpose of this EP is to define custom headers. So, a route owner shall have the choice to define what headers he wants as per his application requirements 
+  than the ones admin has set cluster-wide.
+
+#### Does a custom header value defined in RouteSpec override the route annotation ?
+  Answer: Yes. A value defined in RouteSpec overrides the route annotation.
```

Comment:
Which route annotation are you referring to?

Reply:
So you mean the custom headers specified in the Route spec override _all_ route annotations (except for `haproxy.router.openshift.io/hsts_header`, as the Route spec cannot specify `strict-transport-security`)?

* `haproxy.router.openshift.io/set-forwarded-headers`
* `haproxy.router.openshift.io/h1-adjust-case`
* `router.openshift.io/cookie_name`
* `router.openshift.io/cookie-same-site`

By the way, where this question come up?